### DISCUSSION TITLE
Should we keep the sha256 hash unchanged if the package doesn't change?

### DISCUSSION THREAD

**Discussion by KyleXie** [pyodide/pyodide#4286] [General]
Currently, every time we build pyodide, all the sha256 hashes in `pyodide-lock.json` changes, no matter the package changed or not. 

I looked at the code, and it seems the reason is after we copy package files to the output directory we compress the files, and this causes the hash value changes.

![image](https://github.com/pyodide/pyodide/assets/2870241/4315d525-c54e-4d47-9893-d391bc84d55e)

I was wondering should we keep the hash value if the package has no changes?

**Comment by ryanking13** [comment]
I think the main reason the hash changes is that zipfile contains timestamp metadata. So, it might be possible to maintain the hash value if we set the [`date_time`](https://docs.python.org/3/library/zipfile.html#zipfile.ZipInfo) to some fixed value (someone would need to do some PoC).

@rth @hoodmane Do you think it is worth keeping the hash value but giving up the correct timestamp? I am not sure what would be the side effect of having incorrect timestamp.

**Reply by rth** [reply]
Yes, I would be +1 to fix the timestamp. I just looked at what https://github.com/kushaldas/asaman does and it looks at the [`SOURCE_DATE_EPOCH` env variable](https://github.com/kushaldas/asaman/blob/main/asaman/__init__.py#L19) cf https://reproducible-builds.org/docs/source-date-epoch/. So we could probably also implement that, though I haven't read that page documentation, it has more best practices on ensuring reproducibility.

Also related e.g. https://github.com/pypa/setuptools/issues/2133

**Comment by rth** [comment]
(I'm going to post my message, since I mostly wrote it already, even if it's a bit redundant with the one @ryanking13 posted a few seconds earlier :) )

Interesting observation indeed! 

So as far as I understand, zip files include last modification date timestamps, which will be different each time a build happens. Not just in that `repack_zip_archive` function, but even regular wheels produced with `pypa/build`.

Build reproducibility is not a very easy topic, for context see https://reproducible-builds.org/ and this discussion regarding [reproducible Python wheels](https://discuss.python.org/t/introducing-asaman-a-tool-to-bulid-reproducible-wheels/10932/3).

I'm not sure that fixing those timestamps in zip file, will ensure reproducibility but it would certainly help. At least if we could have the same build with a given Docker image produces wheels with the same checksums, that would be great.

Would you be interested in looking into that @KyleXie ? I think we would need,
 - a function that either sets the last modified dates for all files in a folder to be zipped to some fixed date. Or maybe doing it for a zip directly, but it means doing yet another copy of the zip file.
 - apply it in the build process (probably before creating the final zip).

**Reply by ryanking13** [reply]
I think your comment is much more detailed :)

**Reply by KyleXie** [reply]
Thanks @rth. Let me look at this during the weekend.

**Comment by agriyakhetarpal** [comment]
Adding for completeness: I've recently worked on a change in [`pyodide-build`](https://github.com/pyodide/pyodide-build/) via https://github.com/pyodide/pyodide-build/pull/109, through which we should be able to produce reproducible archives. Please feel free to test it out with an editable installation if there's an issue somewhere; it should be generally available in the next release for `pyodide-build`.

Note that reproducibility is provided on a best-effort basis, as `pyodide-build` as a build frontend cannot govern what the build backend for the package that it's trying to build is doing, and it can't control Emscripten as a compiler (which, based on LLVM, is as reproducible as it can guarantee).
